core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java

```diff
@@ -0,0 +1,320 @@
+        }
+
+        IndexedPriorityQueue<Integer> maxWriters = new IndexedPriorityQueue<>();
+        IndexedPriorityQueue<Integer> minWriters = new IndexedPriorityQueue<>();
+        for (int writerId = 0; writerId < numberOfWriters; writerId++) {
+            long writtenBytes = writerWrittenBytesSinceLastRebalance[writerId];
+            maxWriters.addOrUpdate(writerId, writtenBytes);
+            minWriters.addOrUpdate(writerId, Long.MAX_VALUE - writtenBytes);
+        }
+
+        Set<Integer> rebalancedPartitionIds = new HashSet<>();
+
+        // Find skewed partitions and scale them across multiple writers
+        while (true) {
+            // Find the writer with maximum physical written bytes since last rebalance
+            EntryWithPriority<Integer> maxWriter = maxWriters.pollEntryWithPriority();
+
+            if (maxWriter == null) {
+                break;
+            }
+
+            // Find the skewness against writer with max physical written bytes since last rebalance
+            List<EntryWithPriority<Integer>> minSkewedWriters = findSkewedMinWriters(maxWriter.priority(), minWriters);
+            if (minSkewedWriters.isEmpty()) {
+                break;
+            }
+
+            int maxWriterId = maxWriter.value();
+            int minWriterIdWherePartitionIsScaled = -1;
+            for (EntryWithPriority<Integer> minSkewedWriter : minSkewedWriters) {
+                int minSkewedWriterId = minSkewedWriter.value();
+
+                Optional<IdWithWrittenBytes> rebalancedPartition = rebalancePartition(
+                        minSkewedWriterId,
+                        writerWrittenBytesSinceLastRebalance[maxWriterId],
+                        writersRowCountSinceLastRebalance[maxWriterId],
+                        writerPartitionIdWithRowCount.get(maxWriterId),
+                        rebalancedPartitionIds);
+
+                if (rebalancedPartition.isPresent()) {
+                    IdWithWrittenBytes partitionIdWithWrittenBytes = rebalancedPartition.get();
+                    rebalancedPartitionIds.add(partitionIdWithWrittenBytes.id);
+
+                    long maxWriterPriority = maxWriter.priority() - (partitionIdWithWrittenBytes.writtenBytes / 2);
+                    long minWriterPriority = minSkewedWriter.priority() + (partitionIdWithWrittenBytes.writtenBytes / 2);
+
+                    maxWriters.addOrUpdate(maxWriterId, maxWriterPriority);
+                    maxWriters.addOrUpdate(minSkewedWriterId, minWriterPriority);
+                    minWriters.addOrUpdate(maxWriterId, Long.MAX_VALUE - maxWriterPriority);
+                    minWriters.addOrUpdate(minSkewedWriterId, Long.MAX_VALUE - minWriterPriority);
+
+                    minWriterIdWherePartitionIsScaled = minSkewedWriterId;
+                    break;
+                }
+            }
+
+            // Add all the min skewed writers back to the priority queue except the one where a new partition is rebalanced
```

Comment:
Could you explain why we don't want to add this writer? We updated its statistics.

Reply:
We already added that writer after rebalancing the partition

```
maxWriters.addOrUpdate(maxWriterId, maxWriterPriority);
maxWriters.addOrUpdate(minSkewedWriterId, minWriterPriority);
minWriters.addOrUpdate(maxWriterId, Long.MAX_VALUE - maxWriterPriority);
minWriters.addOrUpdate(minSkewedWriterId, Long.MAX_VALUE - minWriterPriority);                     
```